subsys/bluetooth/mesh/vnd/dm_srv.c

```diff
@@ -0,0 +1,474 @@
+{
+	dm_srv->results.last_entry_idx =
+		(dm_srv->results.last_entry_idx + 1) % dm_srv->results.entry_cnt;
+
+	struct bt_mesh_dm_res_entry *entry = &dm_srv->results.res[dm_srv->results.last_entry_idx];
+
+	dm_srv->results.available_entries =
+		(dm_srv->results.available_entries + 1) >= dm_srv->results.entry_cnt ?
+			dm_srv->results.entry_cnt :
+			(dm_srv->results.available_entries + 1);
+
+	memset(entry, 0, sizeof(struct bt_mesh_dm_res_entry));
+
+	entry->method = result->ranging_mode;
+	entry->addr = dm_srv->target_addr;
+	entry->quality = result->quality;
+	entry->err_occurred = !result->status;
+
+	if (entry->err_occurred) {
+		return;
+	}
+
+	if (entry->method == DM_RANGING_MODE_RTT) {
+		entry->res.rtt = result_convert_cm(result->dist_estimates.rtt.rtt);
+	} else {
+		entry->res.mcpd.best = result_convert_cm(result->dist_estimates.mcpd.best);
+		entry->res.mcpd.ifft = result_convert_cm(result->dist_estimates.mcpd.ifft);
+		entry->res.mcpd.phase_slope =
+			result_convert_cm(result->dist_estimates.mcpd.phase_slope);
+		entry->res.mcpd.rssi_openspace =
+			result_convert_cm(result->dist_estimates.mcpd.rssi_openspace);
+	}
+}
+
+static void cfg_status_send(struct bt_mesh_model *model, struct bt_mesh_msg_ctx *rx_ctx,
+			    uint8_t status)
+{
+	struct bt_mesh_dm_srv *srv = model->user_data;
+
+	BT_MESH_MODEL_BUF_DEFINE(msg, BT_MESH_DM_CLI_OP,
+				 BT_MESH_DM_CONFIG_STATUS_MSG_LEN);
+	bt_mesh_model_msg_init(&msg, BT_MESH_DM_CLI_OP);
+	net_buf_simple_add_u8(&msg, BT_MESH_DM_CONFIG_STATUS_OP);
+	net_buf_simple_add_u8(&msg, status);
+	net_buf_simple_add_u8(&msg, (srv->is_busy << 7) |
+					    (srv->results.available_entries & 0x7F));
+	net_buf_simple_add_u8(&msg, srv->cfg.delay);
+	net_buf_simple_add_u8(&msg, srv->cfg.ttl);
+	net_buf_simple_add_u8(&msg, srv->cfg.timeout);
+
+	(void)bt_mesh_model_send(model, rx_ctx, &msg, NULL, NULL);
+}
+
+static void result_pack(struct bt_mesh_dm_res_entry *entry, struct net_buf_simple *buf)
+{
+	uint8_t temp = (entry->method << 7);
+
+	temp |= ((entry->quality & 0x7) << 4);
+	temp |= ((entry->err_occurred & 0x1) << 3);
+
+	net_buf_simple_add_u8(buf, temp);
+	net_buf_simple_add_le16(buf, entry->addr);
+
+	if (entry->err_occurred) {
+		return;
+	}
+
+	if (entry->method == DM_RANGING_MODE_RTT) {
+		net_buf_simple_add_le16(buf, entry->res.rtt);
+	} else {
+		net_buf_simple_add_le16(buf, entry->res.mcpd.best);
+		net_buf_simple_add_le16(buf, entry->res.mcpd.ifft);
+		net_buf_simple_add_le16(buf, entry->res.mcpd.phase_slope);
+		net_buf_simple_add_le16(buf, entry->res.mcpd.rssi_openspace);
+	}
+}
+
+static void result_status_populate(struct bt_mesh_model *model, struct net_buf_simple *buf,
+				   uint8_t cnt)
+{
+	struct bt_mesh_dm_srv *srv = model->user_data;
+
+	for (int i = 0; i < MIN(cnt, srv->results.available_entries); i++) {
+		struct bt_mesh_dm_res_entry *entry =
+			&srv->results
+				 .res[(srv->results.last_entry_idx + srv->results.entry_cnt - i) %
+				      srv->results.entry_cnt];
```

Comment:
This (and other ring buffer implementation details) should ideally be extracted as functions/macros. Alternatively, look into just using [`ring_buffer.h`](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/sys/ring_buffer.h) (might be too bloated but would avoid reinventing the ~wheel~ ring)